File: integrations/sqlite/typescript/src/approvals.ts
import { wakeRecoverySchedule } from "./recovery-schedule.js";
import type { DatabaseSync } from "node:sqlite";
import { AgentError, ApprovalIntent, ApprovalRequired, copyToolExecutionCheckpoint, copyApprovalDecisionCommand, copyApprovalRecord, jsonIdentityDigest,
  type AgentToolExecutionCheckpoint, type ToolApprovalGateway, type ToolIdempotencyGateway, type ToolDispatchContext, type ApprovalRecord, type ApprovalDecisionCommand, type ApprovalDecisionAudit, type StorageStores } from "purra";
import { storageVersion } from "./approval-format.js";

type Operation<T> = (db: DatabaseSync, scope: string, all: StorageStores, extra: Record<string, any>) => Promise<T>;
export interface ApprovalStorageAccess {
  readonly bindRuntimeClock?: (clock: () => number) => void;
  readonly owner?: () => string | undefined;
  readonly epoch?: () => number | undefined;
  readonly idempotency?: ToolIdempotencyGateway;
  read<T>(operation: (db: DatabaseSync, scope: string) => Promise<T>): Promise<T>;
  write<T>(operation: Operation<T>): Promise<T>;
}
export type ApprovalAuthorizer = (principalId: string, record: ApprovalRecord, command: ApprovalDecisionCommand) => boolean | Promise<boolean>;
export interface ApprovalDecisionReceipt {
  readonly approvalId: string;
  readonly intentDigest: string;
  readonly commandKey: string;
  readonly status: "approved" | "rejected";
  readonly revision: number;
  readonly decidedAtMs: number;
}
function fail(code: string): never { throw new AgentError(code, "Approval operation was rejected"); }
function text(value: string): void {
  if (typeof value !== "string" || !value || value.trim() !== value || [...value].length > 1024) throw new TypeError("Invalid approval text");
}
function integer(value: number): number {
  if (!Number.isSafeInteger(value) || value < 0) throw new TypeError("Invalid approval time");
  return value;
}
function enabled(db: DatabaseSync): void { if (storageVersion(db) !== 5) fail("approval_storage_not_enabled"); }
async function load(db: DatabaseSync, scope: string, id: string): Promise<ApprovalRecord> {
  text(id); enabled(db);
  const row = db.prepare("SELECT run_id,call_id,body FROM purra_approvals WHERE scope=? AND sdk='typescript' AND approval_id=?").get(scope, id);
  if (!row) fail("approval_not_found");
  return decode(id, row);
}
async function decode(id: string, row: Record<string, unknown>): Promise<ApprovalRecord> {
  const record = await copyApprovalRecord(JSON.parse(String(row.body)));
  if (record.approvalId !== id || record.intent.runId !== row.run_id || record.intent.toolCallId !== row.call_id) fail("approval_record_conflict");
  return record;
}
function save(db: DatabaseSync, scope: string, record: ApprovalRecord): void {
  db.prepare("INSERT INTO purra_approvals VALUES(?, 'typescript', ?, ?, ?, ?) ON CONFLICT(scope,sdk,approval_id) DO UPDATE SET body=excluded.body")
    .run(scope, record.approvalId, record.intent.runId, record.intent.toolCallId, JSON.stringify(record));
}
async function runState(db: DatabaseSync, scope: string, all: StorageStores, extra: Record<string, any>, intent: ApprovalRecord["intent"]) {
  const binding = db.prepare("SELECT root_run_id FROM purra_journal_runs WHERE scope=? AND sdk='typescript' AND run_id=?").get(scope, intent.runId);
  if (binding?.root_run_id !== intent.rootRunId) fail("approval_run_conflict");
  const run = await all.runs.get(intent.runId), root = await all.runs.get(intent.rootRunId);
  const canceled = [run, root].some(row => row.status !== "running");
  const deadlines = [run, root].flatMap(row => row.deadlineAt === null ? [] : [Date.parse(row.deadlineAt)]);
  return { canceled, deadline: deadlines.length ? Math.min(...deadlines) : undefined, fingerprint: await jsonIdentityDigest(run.preset) };
}
function receipt(record: ApprovalRecord): ApprovalDecisionReceipt {
  const audit = record.decisionAudit as ApprovalDecisionAudit;
  return Object.freeze({ approvalId: record.approvalId, intentDigest: record.intentDigest, commandKey: audit.command.commandKey,
    status: audit.command.decision === "approve" ? "approved" : "rejected", revision: audit.revision, decidedAtMs: audit.decidedAtMs });
}

/** Host-authorized decisions and atomic durable waits. Dispatch is revalidated separately. */
export class SqliteApprovalStore {
  constructor(private readonly access: ApprovalStorageAccess, private readonly authorize: ApprovalAuthorizer, private readonly clock: () => number = Date.now) {
    if (typeof authorize !== "function") throw new TypeError("Approval authorizer is required");
  }

  async create(intent: ApprovalIntent, options: { expiresAtMs: number }): Promise<ApprovalRecord> {
    if (!(intent instanceof ApprovalIntent)) throw new TypeError("Approval intent is required");
    const requestedExpiry = integer(options.expiresAtMs);
    return this.access.write((db, scope, all, extra) => this.#create(db, scope, all, extra, intent, requestedExpiry));
  }

  async #create(db: DatabaseSync, scope: string, all: StorageStores, extra: Record<string, any>, intent: ApprovalIntent, requestedExpiry: number): Promise<ApprovalRecord> {
    const id = await jsonIdentityDigest({ profile: "purra.approval-key/v1", runId: intent.value.runId, toolCallId: intent.value.toolCallId });
    enabled(db);
    const state = await runState(db, scope, all, extra, intent.value);
    if (state.canceled) fail("approval_run_terminal");
    if (state.fingerprint !== intent.value.presetFingerprint) fail("approval_configuration_mismatch");
    const expiresAtMs = Math.min(requestedExpiry, state.deadline ?? requestedExpiry);
    if (db.prepare("SELECT 1 FROM purra_approvals WHERE scope=? AND sdk='typescript' AND approval_id=?").get(scope, id)) {
      const record = await load(db, scope, id);
      if (record.intentDigest !== intent.digest || record.expiresAtMs !== expiresAtMs) fail("approval_intent_conflict");
      return record;
    }
    const now = integer(this.clock());
    if (now >= expiresAtMs) fail("approval_expired");
    const record = await copyApprovalRecord({ approvalId: id, intent: intent.value, intentDigest: intent.digest,
      revision: 1, status: "pending", createdAtMs: now, expiresAtMs, decisionAudit: {} });
    save(db, scope, record);
    return record;
  }

  async prepare(checkpoint: AgentToolExecutionCheckpoint, intent: ApprovalIntent, options: { expiresAtMs: number }): Promise<void> {
    const copied = copyToolExecutionCheckpoint(checkpoint), expiry = integer(options.expiresAtMs);
    if (!(intent instanceof ApprovalIntent)) throw new TypeError("Approval intent is required");
    const call = copied.assistant.toolCalls![0]!;
    if (copied.runId !== intent.value.runId || call.id !== intent.value.toolCallId || call.name !== intent.value.toolName
      || await jsonIdentityDigest(call.arguments) !== await jsonIdentityDigest(intent.value.arguments)) fail("approval_intent_conflict");
    const outcome = await this.access.write(async (db, scope, all, extra) => {
      requireApprovalOwner(extra, copied.runId, this.access.owner?.(), this.access.epoch?.());
      let record = await this.#create(db, scope, all, extra, intent, expiry);
      record = (await this.#refresh(db, scope, all, extra, record, integer(this.clock()))).record;
      await all.runs.saveToolExecutionCheckpoint(copied.runId, copied);
      await all.runs.appendEvent(copied.runId, { sourceKey: `approval-required:${record.approvalId}`, kind: "approval.required",
        channel: "lifecycle", visibility: "private", payload: { approvalId: record.approvalId } });
      const completed = Object.values(extra.tools).some((entry: any) => entry.approvalId === record.approvalId && entry.intentDigest === record.intentDigest
          && entry.runId === record.intent.runId && entry.callId === record.intent.toolCallId && entry.state === "complete");
      return { record, completed };
    });
    if (outcome.completed) return;
    if (outcome.record.status === "pending") throw new ApprovalRequired(copied.runId, outcome.record.approvalId);
    if (outcome.record.status !== "approved") fail(`approval_${outcome.record.status}`);
  }

  gateway(): ToolApprovalGateway {
    if (this.access.idempotency === undefined || this.access.owner === undefined) fail("approval_runtime_required");
    this.access.bindRuntimeClock?.(this.clock);
    return Object.freeze({ requiresDurableIdempotency: true, idempotencyGateway: this.access.idempotency,
      request: async (request: Parameters<ToolApprovalGateway["request"]>[0]) => {
        if (request.dispatch === undefined) fail("approval_run_conflict");
        const outcome = await this.access.write((db, scope, all, extra) => checkApprovalDispatch(
          db, scope, all, extra, request.dispatch!, this.access.owner!(), integer(this.clock()), this.access.epoch?.()));
        if (outcome.error !== undefined) fail(outcome.error);
        return "approved" as const;
      },
    });
  }

  async get(approvalId: string): Promise<ApprovalRecord> {
    return this.access.read((db, scope) => load(db, scope, approvalId));
  }

  async listPending(options: { runId?: string } = {}): Promise<readonly ApprovalRecord[]> {
    const runId = options.runId;
    if (runId !== undefined) text(runId);
    return this.access.read(async (db, scope) => {
      enabled(db);
      const query = db.prepare("SELECT approval_id,run_id,call_id,body FROM purra_approvals WHERE scope=? AND sdk='typescript'"
        + (runId === undefined ? "" : " AND run_id=?") + " ORDER BY approval_id");
      const rows = runId === undefined ? query.all(scope) : query.all(scope, runId);
      const records = await Promise.all(rows.map(row => decode(String(row.approval_id), row)));
      return Object.freeze(records.filter(record => ["pending", "approved"].includes(record.status)));
    });
  }

  async #refresh(db: DatabaseSync, scope: string, all: StorageStores, extra: Record<string, any>, record: ApprovalRecord, now: number) {
    const state = await runState(db, scope, all, extra, record.intent);
    const status = state.canceled ? "canceled" : now >= Math.min(record.expiresAtMs, state.deadline ?? record.expiresAtMs) ? "expired" : undefined;
    if (status !== undefined && ["pending", "approved"].includes(record.status)) {
      record = await copyApprovalRecord({ ...record, status, revision: record.revision + 1 });
      save(db, scope, record);
    }
    return { record, fingerprint: state.fingerprint };
  }

  async refresh(approvalId: string): Promise<ApprovalRecord> {
    return this.access.write(async (db, scope, all, extra) =>
      (await this.#refresh(db, scope, all, extra, await load(db, scope, approvalId), integer(this.clock()))).record);
  }

  async decide(input: ApprovalDecisionCommand, options: { principalId: string }): Promise<ApprovalDecisionReceipt> {
    const command = copyApprovalDecisionCommand(input), principalId = options.principalId;
    text(principalId);
    const observed = await this.get(command.approvalId);
    let allowed: boolean;
    try { allowed = await this.authorize(principalId, observed, command); }
    catch { return fail("approval_authorization_failed"); }
    if (allowed !== true) fail("approval_authorization_denied");
    const result = await this.access.write(async (db, scope, all, extra) => {
      let record = await load(db, scope, command.approvalId);
      const audit = record.decisionAudit as ApprovalDecisionAudit;
      if (audit.command?.commandKey === command.commandKey) {
        if (audit.principalId !== principalId || await jsonIdentityDigest(audit.command) !== await jsonIdentityDigest(command)) fail("approval_command_conflict");
        return { receipt: receipt(record) };
      }
      if (record.revision !== command.expectedRevision || record.intentDigest !== command.intentDigest || record.status !== "pending") fail("approval_revision_conflict");
      const now = integer(this.clock());
      const refreshed = await this.#refresh(db, scope, all, extra, record, now);
      record = refreshed.record;
      if (["expired", "canceled"].includes(record.status)) return { error: "approval_" + record.status };
      if (refreshed.fingerprint !== record.intent.presetFingerprint) return { error: "approval_configuration_mismatch" };
      record = await copyApprovalRecord({ ...record, status: command.decision === "approve" ? "approved" : "rejected", revision: record.revision + 1,
        decisionAudit: { command, principalId, decidedAtMs: now, revision: record.revision + 1 } });
      save(db, scope, record);
      wakeRecoverySchedule(extra, record.intent.runId, true);
      return { receipt: receipt(record) };
    });
    if (result.error) fail(result.error);
    return result.receipt!;
  }
}

export function requireApprovalOwner(extra: Record<string, any>, runId: string, owner: string | undefined, epoch?: number) {
  const lease = extra.leases[runId];
  if (owner === undefined || lease?.owner !== owner || lease.expires <= Date.now()
    || !Number.isSafeInteger(lease.epoch) || lease.epoch < 1 || epoch !== undefined && epoch !== lease.epoch) fail("run_lease_lost");
  return lease;
}

export async function checkApprovalDispatch(db: DatabaseSync, scope: string, all: StorageStores, extra: Record<string, any>,
  dispatch: ToolDispatchContext, owner: string | undefined, now: number, epoch?: number) {
  enabled(db);
  requireApprovalOwner(extra, dispatch.runId, owner, epoch);
  const row = db.prepare("SELECT approval_id FROM purra_approvals WHERE scope=? AND sdk='typescript' AND run_id=? AND call_id=?")
    .get(scope, dispatch.runId, dispatch.call.id);
  if (!row) fail("approval_not_found");
  let record = await load(db, scope, String(row.approval_id));
  if (dispatch.approvalBinding !== undefined && Object.entries(dispatch.approvalBinding).some(([key, value]) =>
    record.intent[key as keyof typeof record.intent] !== value)) fail("approval_intent_conflict");
  const run = await all.runs.get(dispatch.runId), checkpoint = run.toolExecutionCheckpoint;
  if (checkpoint === undefined || await jsonIdentityDigest(checkpoint.assistant.toolCalls![0]!) !== await jsonIdentityDigest(dispatch.call)
    || record.intent.toolName !== dispatch.call.name
    || await jsonIdentityDigest(record.intent.arguments) !== await jsonIdentityDigest(dispatch.call.arguments)) fail("approval_intent_conflict");
  if (!all.runs.hasSettledToolInvocation(dispatch.runId, checkpoint.invocationId)) fail("approval_invocation_unsettled");
  const state = await runState(db, scope, all, extra, record.intent);
  if (state.canceled) return { record, error: "approval_canceled" };
  if (state.fingerprint !== record.intent.presetFingerprint) return { record, error: "approval_configuration_mismatch" };
  const completed = Object.values(extra.tools).some((entry: any) => entry.approvalId === record.approvalId && entry.intentDigest === record.intentDigest
          && entry.runId === record.intent.runId && entry.callId === record.intent.toolCallId && entry.state === "complete");
  if (completed) return { record };
  if (now >= Math.min(record.expiresAtMs, state.deadline ?? record.expiresAtMs) && ["pending", "approved"].includes(record.status)) {
    record = await copyApprovalRecord({ ...record, status: "expired", revision: record.revision + 1 });
    save(db, scope, record);
  }
  return { record, ...(record.status === "approved" ? {} : { error: `approval_${record.status}` }) };
}


/** Bind host reconciliation to the persisted approval; it grants no new dispatch. */
export async function checkApprovalReconciliation(db: DatabaseSync, scope: string, claim: Record<string, any>): Promise<void> {
  enabled(db);
  if (typeof claim.approvalId !== "string" || !claim.approvalId) fail("approval_claim_conflict");
  const row = db.prepare("SELECT approval_id,run_id,call_id,body FROM purra_approvals WHERE scope=? AND sdk='typescript' AND approval_id=?")
    .get(scope, claim.approvalId);
  if (!row || row.run_id !== claim.runId || row.call_id !== claim.callId) fail("approval_claim_conflict");
  const record = await decode(String(row.approval_id), row);
  const audit = record.decisionAudit as Partial<ApprovalDecisionAudit>;
  if (claim.state !== "claimed" || claim.intentDigest !== record.intentDigest
    || claim.approvalRevision !== audit.revision || audit.command?.decision !== "approve") fail("approval_claim_conflict");
}

/** Snapshot-only observations. Never refreshes approval state or acquires ownership. */
export async function inspectApprovalState(db: DatabaseSync, scope: string, saved: Awaited<ReturnType<StorageStores["runs"]["get"]>>, extra: Record<string, any>, now: number) {
  if (storageVersion(db) !== 5) return {};
  const rows = db.prepare("SELECT approval_id FROM purra_approvals WHERE scope=? AND sdk='typescript' AND run_id=?").all(scope, saved.runId);
  const records = await Promise.all(rows.map(row => load(db, scope, String(row.approval_id))));
  const unknown = Object.values(extra.tools).filter((entry: any) => entry.state === "claimed" && entry.runId === saved.runId
    && records.some(record => record.approvalId === entry.approvalId && record.intentDigest === entry.intentDigest && record.intent.toolCallId === entry.callId)).length;
  const observation: import("purra").RecoveryObservations = {approvalState: records.length ? "unknown" : "none", approvalRecords: records.length,
    approvalUnknownReceipts: unknown, approvalCheckpointIntent: "unknown", approvalReceipt: "unknown"};
  const call = saved.toolExecutionCheckpoint?.assistant.toolCalls?.[0];
  if (call === undefined) return observation;
  const record = records.find(record => record.intent.toolCallId === call.id);
  if (record === undefined) return {...observation, approvalState: "missing" as const, approvalReceipt: "absent" as const};
  const matches = record.intent.toolName === call.name && await jsonIdentityDigest(record.intent.arguments) === await jsonIdentityDigest(call.arguments)
    && record.intent.presetFingerprint === await jsonIdentityDigest(saved.preset);
  const complete = matches && Object.values(extra.tools).some((entry: any) => entry.state === "complete" && entry.runId === saved.runId
    && entry.callId === call.id && entry.approvalId === record.approvalId && entry.intentDigest === record.intentDigest
    && ["not_started", "committed"].includes(entry.result?.effectState)
    && entry.approvalRevision === record.decisionAudit?.revision && typeof entry.leaseOwnerId === "string" && entry.leaseOwnerId.length > 0
    && Number.isSafeInteger(entry.leaseEpoch) && entry.leaseEpoch > 0);
  let state = record.status;
  if (state === "pending" || state === "approved") {
    if (saved.status !== "running") state = "canceled";
    else if (now >= Math.min(record.expiresAtMs, saved.deadlineAt === null ? record.expiresAtMs : Date.parse(saved.deadlineAt))) state = "expired";
  }
  return {...observation, approvalState: state, approvalCheckpointIntent: matches ? "matched" as const : "mismatch" as const,
    approvalReceipt: complete ? "complete" as const : "absent" as const};
}

File: integrations/sqlite/python/src/purra_sqlite/approvals.py
"""Transactional approval decisions; runtime suspension/dispatch is separate."""
from dataclasses import replace
import inspect
import json
import time

from purra.approvals import ApprovalIntent, ApprovalRecord, ApprovalDecisionCommand
from purra.errors import ContractViolationError
from purra.json_values import freeze_json_mapping, thaw_json_mapping
from purra.structured import json_identity_digest
from .approval_format import storage_version


def _fail(code):
    raise ContractViolationError("Approval operation was rejected", code=code)


def _text(value, name):
    if not isinstance(value, str) or not value or value.strip() != value or len(value) > 1024:
        raise ValueError(f"invalid approval {name}")


def _integer(value, name):
    if type(value) is not int or not 0 <= value <= 9007199254740991:
        raise ValueError(f"invalid approval {name}")
    return value


class SqliteApprovalStore:
    """Host-only intent/decision storage. Approved records are not execution permits.

    authorize(principal_id, record, command) must return literal True. Obtain
    principal_id from authenticated host context, never from model/tool data.
    """
    def __init__(self, storage, *, authorize, clock_ms=None):
        if not callable(authorize):
            raise TypeError("approval authorizer is required")
        self._storage, self._authorize = storage, authorize
        self._clock = clock_ms or (lambda: int(time.time() * 1000))

    def _now(self):
        return _integer(self._clock(), "clock")

    def _enabled(self):
        if storage_version(self._storage._db) != 5:
            _fail("approval_storage_not_enabled")

    def _load(self, approval_id):
        _text(approval_id, "id")
        self._enabled()
        row = self._storage._db.execute(
            "SELECT run_id,call_id,body FROM purra_approvals WHERE scope=? AND sdk='python' AND approval_id=?",
            (self._storage.scope, approval_id),
        ).fetchone()
        if row is None:
            _fail("approval_not_found")
        return self._decode(approval_id, row)

    @staticmethod
    def _decode(approval_id, row):
        record = ApprovalRecord.from_mapping(json.loads(row[2]))
        if record.approval_id != approval_id or (record.intent.run_id, record.intent.tool_call_id) != row[:2]:
            _fail("approval_record_conflict")
        return record

    def _save(self, record):
        self._storage._db.execute(
            "INSERT INTO purra_approvals VALUES(?, 'python', ?, ?, ?, ?) "
            "ON CONFLICT(scope,sdk,approval_id) DO UPDATE SET body=excluded.body",
            (self._storage.scope, record.approval_id, record.intent.run_id, record.intent.tool_call_id,
             json.dumps(record.to_mapping(), ensure_ascii=False, separators=(",", ":"))),
        )

    async def _run_state(self, session, intent):
        info = session.get_run_info(intent.run_id)
        if info is None or info.root_run_id != intent.root_run_id:
            _fail("approval_run_conflict")
        run = await session.runs.get(intent.run_id)
        root = await session.runs.get(intent.root_run_id)
        canceled = any(row.status.value != "running" for row in (run, root)) or any(
            lease is not None and lease.cancellation_requested_at_ms is not None
            for lease in (session.leases.get(intent.run_id), session.leases.get(intent.root_run_id))
        )
        deadlines = [row.deadline_at_ms for row in (run, root) if row.deadline_at_ms is not None]
        return canceled, min(deadlines) if deadlines else None, json_identity_digest(run.agent_preset_snapshot)

    async def create(self, intent: ApprovalIntent, *, expires_at_ms: int):
        if not isinstance(intent, ApprovalIntent):
            raise TypeError("approval intent is required")
        _integer(expires_at_ms, "expiry")
        identifier = json_identity_digest({"profile": "purra.approval-key/v1", "runId": intent.run_id, "toolCallId": intent.tool_call_id})
        async with self._storage._transaction(with_journal=False) as session:
            return await self._create(session, intent, expires_at_ms, identifier)

    async def _create(self, session, intent, expires_at_ms, identifier):
        self._enabled()
        canceled, deadline, fingerprint = await self._run_state(session, intent)
        if canceled:
            _fail("approval_run_terminal")
        if fingerprint != intent.preset_fingerprint:
            _fail("approval_configuration_mismatch")
        expiry = min(expires_at_ms, deadline) if deadline is not None else expires_at_ms
        existing = self._storage._db.execute(
            "SELECT 1 FROM purra_approvals WHERE scope=? AND sdk='python' AND approval_id=?",
            (self._storage.scope, identifier),
        ).fetchone()
        if existing:
            record = self._load(identifier)
            if record.intent.digest != intent.digest or record.expires_at_ms != expiry:
                _fail("approval_intent_conflict")
            return record
        now = self._now()
        if now >= expiry:
            _fail("approval_expired")
        record = ApprovalRecord(identifier, intent, 1, "pending", now, expiry)
        self._save(record)
        return record

    async def prepare(self, checkpoint, intent, *, expires_at_ms):
        """Commit the intent and real tool-ready cursor together, then suspend.

        Called only by a leased runtime boundary. Reconstruct intent from current
        host bindings on every resume; no record itself grants dispatch authority.
        """
        from purra.agent_execution_checkpoint import AgentToolExecutionCheckpoint
        from purra.approvals import ApprovalRequired
        from purra.ports import RunCommit
        from purra.events import AgentEvent
        if not isinstance(checkpoint, AgentToolExecutionCheckpoint) or not isinstance(intent, ApprovalIntent):
            raise TypeError("tool-ready checkpoint and intent required")
        _integer(expires_at_ms, "expiry")
        call = checkpoint.assistant.tool_calls[0]
        if (checkpoint.run_id != intent.run_id or call.id != intent.tool_call_id or call.name != intent.tool_name
                or json_identity_digest(json.loads(call.arguments_json)) != json_identity_digest(intent.arguments)):
            _fail("approval_checkpoint_conflict")
        identifier = json_identity_digest({"profile": "purra.approval-key/v1", "runId": intent.run_id, "toolCallId": intent.tool_call_id})
        async with self._storage._transaction() as session:
            self._require_owner(session, intent.run_id)
            if not session.has_settled_tool_invocation(intent.run_id, checkpoint.invocation_id, checkpoint.model_budget_key):
                _fail("approval_invocation_unsettled")
            record = await self._create(session, intent, expires_at_ms, identifier)
            current = (await session.runs.get(intent.run_id)).execution_checkpoint
            if current != checkpoint:
                await session.runs.commit(intent.run_id, RunCommit(execution_checkpoint=checkpoint,
                    events=(AgentEvent("agent.execution_checkpointed", {"schemaVersion": 3, "phase": "tool_ready", "nextRound": checkpoint.next_round}, intent.run_id),)))
                if record.status == "pending":
                    from datetime import datetime, timezone
                    from purra.output import AgentOutputEventDraft
                    await session.outputs.append_event(AgentOutputEventDraft(
                        run_id=intent.run_id, turn_id=None, output_stream_id=None, invocation_id=checkpoint.invocation_id,
                        source_event_key=f"approval:{record.approval_id}:required:{record.revision}",
                        source="runtime", kind="runtime.event", channel="lifecycle", visibility="private",
                        payload={"type": "approval.required", "status": "pending", "revision": record.revision},
                        occurred_at=datetime.now(timezone.utc)))
            record, _ = await self._refresh(session, record, self._now())
            completed = self._completed_receipt(session, record, call)
        if completed:
            return record
        if record.status == "pending":
            raise ApprovalRequired(record.intent.run_id, record.approval_id)
        if record.status != "approved":
            _fail("approval_" + record.status)
        return record

    def _require_owner(self, session, run_id):
        from purra.execution.ownership import execution_owner, execution_claim
        claim = execution_claim.get()
        lease = session.leases.get(run_id)
        if (claim is None or claim[0] != run_id or lease is None or execution_owner.get() is None
                or lease.owner_id != execution_owner.get() or claim[2] != lease.attempt
                or (lease.expires_at_ms or 0) <= int(time.time() * 1000)):
            _fail("agent_run_lease_lost")

    def gateway(self):
        """Bind only alongside prepare and this adapter's idempotency gateway."""
        return _DurableApprovalGateway(self)

    @staticmethod
    def _completed_receipt(session, record, call):
        receipt = session.get_tool_receipt((record.intent.run_id, call.id))
        if receipt is None:
            return False
        association = session.extra.get("approvalExecutions", {}).get(record.approval_id)
        if (receipt[0] != call or receipt[1].effect_state.value == "unknown" or not isinstance(association, dict)
                or set(association) != {"intentDigest", "state", "effectState", "approvalRevision", "leaseOwnerId", "leaseEpoch"}
                or association["intentDigest"] != record.intent.digest or association["state"] != "complete"
                or association["effectState"] != receipt[1].effect_state.value
                or association["approvalRevision"] != record.decision_audit.get("revision")
                or not isinstance(association["leaseOwnerId"], str) or not association["leaseOwnerId"]
                or type(association["leaseEpoch"]) is not int or association["leaseEpoch"] < 1):
            _fail("approval_claim_conflict")
        return True

    async def _dispatch_record(self, session, run_id, call):
        self._enabled()
        from purra.agent_execution_checkpoint import AgentToolExecutionCheckpoint
        self._require_owner(session, run_id)
        identifier = json_identity_digest({"profile": "purra.approval-key/v1", "runId": run_id, "toolCallId": call.id})
        record = self._load(identifier)
        intent = record.intent
        checkpoint = (await session.runs.get(run_id)).execution_checkpoint
        if (not isinstance(checkpoint, AgentToolExecutionCheckpoint) or checkpoint.assistant.tool_calls != (call,)
                or call.name != intent.tool_name
                or json_identity_digest(json.loads(call.arguments_json)) != json_identity_digest(intent.arguments)):
            _fail("approval_checkpoint_conflict")
        if not session.has_settled_tool_invocation(run_id, checkpoint.invocation_id, checkpoint.model_budget_key):
            _fail("approval_invocation_unsettled")
        if self._completed_receipt(session, record, call):
            return record
        canceled, deadline, fingerprint = await self._run_state(session, intent)
        if canceled:
            _fail("approval_canceled")
        if record.status != "approved":
            _fail("approval_" + record.status)
        if self._now() >= min(record.expires_at_ms, deadline if deadline is not None else record.expires_at_ms):
            _fail("approval_expired")
        if fingerprint != intent.preset_fingerprint:
            _fail("approval_configuration_mismatch")
        return record

    async def get(self, approval_id):
        async with self._storage._connection(read_only=True):
            return self._load(approval_id)

    async def list_pending(self, *, run_id=None):
        if run_id is not None:
            _text(run_id, "Run id")
        async with self._storage._connection(read_only=True):
            self._enabled()
            rows = self._storage._db.execute(
                "SELECT approval_id,run_id,call_id,body FROM purra_approvals WHERE scope=? AND sdk='python'"
                + (" AND run_id=?" if run_id is not None else "") + " ORDER BY approval_id",
                (self._storage.scope, run_id) if run_id is not None else (self._storage.scope,),
            ).fetchall()
            return tuple(record for row in rows if (record := self._decode(row[0], row[1:])).status in {"pending", "approved"})

    async def _refresh(self, session, record, now):
        canceled, deadline, fingerprint = await self._run_state(session, record.intent)
        status = "canceled" if canceled else "expired" if now >= min(record.expires_at_ms, deadline if deadline is not None else record.expires_at_ms) else None
        if status is not None and record.status in {"pending", "approved"}:
            record = replace(record, status=status, revision=record.revision + 1)
            self._save(record)
        return record, fingerprint

    async def refresh(self, approval_id):
        """Persist only expiry/canonical Run cancellation, never a human decision."""
        async with self._storage._transaction(with_journal=False) as session:
            record, _ = await self._refresh(session, self._load(approval_id), self._now())
            return record

    async def decide(self, command: ApprovalDecisionCommand, *, principal_id: str):
        if not isinstance(command, ApprovalDecisionCommand):
            raise TypeError("approval decision command is required")
        _text(principal_id, "principal")
        observed = await self.get(command.approval_id)
        try:
            allowed = self._authorize(principal_id, observed, command)
            if inspect.isawaitable(allowed):
                allowed = await allowed
        except Exception:
            _fail("approval_authorization_failed")
        if allowed is not True:
            _fail("approval_authorization_denied")
        error = None
        async with self._storage._transaction(with_journal=False) as session:
            record = self._load(command.approval_id)
            audit = record.decision_audit
            if audit and audit["command"]["commandKey"] == command.command_key:
                if audit["principalId"] != principal_id or thaw_json_mapping(audit["command"]) != command.to_mapping():
                    _fail("approval_command_conflict")
                return _receipt(record)
            if record.revision != command.expected_revision or record.intent.digest != command.intent_digest or record.status != "pending":
                _fail("approval_revision_conflict")
            now = self._now()
            record, fingerprint = await self._refresh(session, record, now)
            if record.status in {"expired", "canceled"}:
                error = "approval_" + record.status
            elif fingerprint != record.intent.preset_fingerprint:
                error = "approval_configuration_mismatch"
            else:
                record = replace(record, status="approved" if command.decision == "approve" else "rejected",
                    revision=record.revision + 1, decision_audit={"command": command.to_mapping(),
                        "principalId": principal_id, "decidedAtMs": now, "revision": record.revision + 1})
                self._save(record)
        if error:
            _fail(error)
        return _receipt(record)


def _receipt(record):
    audit = record.decision_audit
    return freeze_json_mapping({"approvalId": record.approval_id, "intentDigest": record.intent.digest,
        "commandKey": audit["command"]["commandKey"], "status": "approved" if audit["command"]["decision"] == "approve" else "rejected",
        "revision": audit["revision"], "decidedAtMs": audit["decidedAtMs"]})


class _DurableApprovalGateway:
    requires_durable_idempotency = True

    @property
    def idempotency_gateway(self):
        return self._store._storage.idempotency

    def __init__(self, store):
        self._store = store

    async def request(self, run_id, approval, event_sink, signal=None):
        from purra.contracts import ApprovalResult, ApprovalStatus
        if signal is not None and signal.is_set():
            return ApprovalResult(None, ApprovalStatus.CANCELED)
        async with self._store._storage._transaction(with_journal=False) as session:
            record = await self._store._dispatch_record(session, run_id, approval.tool_call)
            if approval.binding is not None:
                intent = record.intent.to_mapping()
                if any(intent[key] != value for key, value in approval.binding.items()):
                    _fail("approval_intent_conflict")
        return ApprovalResult(record.approval_id, ApprovalStatus.APPROVED)

    async def resolve(self, run_id, approval_id, decision):
        return None

    async def cancel_pending(self, run_id):
        # Suspending a durable Run must not cancel its persisted intent.
        return 0


def inspect_approval_state(storage, session, saved, now):
    """Read committed approval evidence without refreshing or authorizing a record."""
    from purra.agent_execution_checkpoint import AgentToolExecutionCheckpoint
    if storage_version(storage._db) != 5:
        return {}
    rows = storage._db.execute("SELECT approval_id,call_id,body FROM purra_approvals WHERE scope=? AND sdk='python' AND run_id=?",
        (storage.scope, saved.run_id)).fetchall()
    records = []
    for identifier, call_id, body in rows:
        record = ApprovalRecord.from_mapping(json.loads(body))
        if record.approval_id != identifier or record.intent.run_id != saved.run_id or record.intent.tool_call_id != call_id:
            _fail("approval_record_conflict")
        records.append(record)
    observation = {"approvalState": "unknown" if records else "none", "approvalRecords": len(records),
        "approvalUnknownReceipts": sum((saved.run_id, record.intent.tool_call_id) in session.claims for record in records),
        "approvalCheckpointIntent": "unknown", "approvalReceipt": "unknown"}
    checkpoint = saved.execution_checkpoint
    if not isinstance(checkpoint, AgentToolExecutionCheckpoint):
        return observation
    call = checkpoint.assistant.tool_calls[0]
    record = next((record for record in records if record.intent.tool_call_id == call.id), None)
    if record is None:
        observation.update(approvalState="missing", approvalReceipt="absent")
        return observation
    matches = (record.intent.tool_name == call.name and json_identity_digest(record.intent.arguments) == json_identity_digest(json.loads(call.arguments_json))
        and record.intent.preset_fingerprint == json_identity_digest(saved.agent_preset_snapshot))
    complete = matches and SqliteApprovalStore._completed_receipt(session, record, call)
    state = record.status
    if state in {"pending", "approved"}:
        if saved.status.value != "running": state = "canceled"
        elif now >= min(record.expires_at_ms, saved.deadline_at_ms if saved.deadline_at_ms is not None else record.expires_at_ms): state = "expired"
    observation.update(approvalState=state, approvalCheckpointIntent="matched" if matches else "mismatch", approvalReceipt="complete" if complete else "absent")
    return observation
